{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "69d0f591",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pyensembl\n",
    "from IPython.display import clear_output, display\n",
    "import torch\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e08ffd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import get_control_file as gcf\n",
    "import position_encoding as pos\n",
    "import pathway_encoding as path\n",
    "from model import TranscriptomePredictor, PerturbationDataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "80f9ed17",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.read_h5ad('../vcc_data/adata_Training.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "eebb6785",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /Users/steveyin/Library/Caches/pyensembl/GRCh38/ensembl109/Homo_sapiens.GRCh38.cdna.all.fa.gz.pickle\n",
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /Users/steveyin/Library/Caches/pyensembl/GRCh38/ensembl109/Homo_sapiens.GRCh38.ncrna.fa.gz.pickle\n",
      "INFO:pyensembl.sequence_data:Loaded sequence dictionary from /Users/steveyin/Library/Caches/pyensembl/GRCh38/ensembl109/Homo_sapiens.GRCh38.pep.all.fa.gz.pickle\n"
     ]
    }
   ],
   "source": [
    "ensembl_data = pyensembl.EnsemblRelease(109)\n",
    "ensembl_data.download()\n",
    "ensembl_data.index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3f2673bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#adata.obs['perturbation_idx'] = np.random.randint(0, CONFIG[\"n_perturbations\"], size=adata.n_obs)\n",
    "gene_names = pd.read_csv('../vcc_data/gene_names.csv', header = None)\n",
    "gene_names = gene_names.iloc[:, 0].tolist()\n",
    "\n",
    "\n",
    "CONFIG = {\n",
    "    \"n_genes\": len(gene_names),\n",
    "    \"n_perturbations\": len(adata.obs['target_gene'].unique().tolist()),\n",
    "    \"n_chromosomes\": 25,\n",
    "\n",
    "    \"perturbation_dim\": 256,      # Condition embedding\n",
    "    \"chrom_embedding_dim\": 16,     # Learned in-model for chromosome identity\n",
    "    \"locus_fourier_features\": 8,   # Number of Fourier frequency pairs (2*F)\n",
    "    \"pathway_dim\": 50,             # From pre-trained Autoencoder(based on hallmark MSigDB)\n",
    "    \"gene_identity_dim\": 189,       # Main learnable gene embedding\n",
    "\n",
    "    # Backbone dims\n",
    "    \"d_model\": 512,                # Mamba hidden size\n",
    "    \"mamba_layers\": 4,\n",
    "\n",
    "    # Head\n",
    "    \"prediction_head\": \"probabilistic\", # \"linear\" | \"probabilistic\"\n",
    "\n",
    "    # Training\n",
    "    \"batch_size\": 64,\n",
    "    \"learning_rate\": 5e-5, # Lowered LR for AdamW stability\n",
    "    \"epochs\": 100,\n",
    "}\n",
    "\n",
    "# Derived dimensions for clarity\n",
    "# Positional encoding dim = chromosome embedding + MLP output from Fourier features\n",
    "POS_DIM = CONFIG[\"chrom_embedding_dim\"] + CONFIG[\"chrom_embedding_dim\"]\n",
    "GENE_FEAT_DIM = CONFIG[\"gene_identity_dim\"] + CONFIG[\"pathway_dim\"] + POS_DIM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cc4e1968",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique_perturbations = adata.obs['target_gene'].unique().tolist()\n",
    "perturbation_to_idx_map = {name: i for i, name in enumerate(unique_perturbations)}\n",
    "adata.obs['perturbation_idx'] = adata.obs['target_gene'].map(perturbation_to_idx_map)\n",
    "control_adata = gcf.get_control_data(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "993064a2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- Preparing positional indices using pyensembl ---\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Fetching gene positions: 100%|██████████| 18080/18080 [00:00<00:00, 60112.75it/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "SUCCESS: Generated positional tensors with shapes:\n",
      "Chromosome Indices (chr_idx): torch.Size([18080])\n",
      "Normalized Locus (locus_norm): torch.Size([18080, 1])\n",
      "Locus Fourier Features (locus_fourier): torch.Size([18080, 16])\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "chr_idx, locus_norm, locus_fourier = pos.precompute_positional_indices(ensembl_data, gene_names, CONFIG)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "65368b05",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--- Precomputing pathway features on control data ---\n",
      "AE epoch 10/100 | recon MSE: 78.6051\n",
      "AE epoch 20/100 | recon MSE: 50.8467\n",
      "AE epoch 30/100 | recon MSE: 25.8626\n",
      "AE epoch 40/100 | recon MSE: 13.6297\n",
      "AE epoch 50/100 | recon MSE: 9.4090\n",
      "AE epoch 60/100 | recon MSE: 7.9185\n",
      "AE epoch 70/100 | recon MSE: 7.3963\n",
      "AE epoch 80/100 | recon MSE: 7.2014\n",
      "AE epoch 90/100 | recon MSE: 7.1255\n",
      "AE epoch 100/100 | recon MSE: 7.0952\n",
      "Generated pathway_features shape: torch.Size([18080, 50])\n"
     ]
    }
   ],
   "source": [
    "pathway_feats = path.precompute_pathway_features(control_adata, CONFIG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7dfb9221",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[-4.0940959e-02, -1.7496973e-02,  3.9910398e-02, ...,\n",
       "         4.0042114e-02, -1.0482197e-02,  2.9803514e-03],\n",
       "       [ 4.3170462e+00,  5.3395712e-01,  3.0772862e+00, ...,\n",
       "         2.6977115e+00, -3.4365354e+00, -3.8631627e-01],\n",
       "       [ 7.8416473e-01,  1.3108829e-01,  5.3170836e-01, ...,\n",
       "         6.6698223e-01, -5.0516069e-01, -7.5793922e-02],\n",
       "       ...,\n",
       "       [ 2.3736744e-01,  9.1434773e-03,  2.5253361e-01, ...,\n",
       "         2.5257570e-01, -1.8625672e-01, -7.2753406e-03],\n",
       "       [ 8.8144159e+00,  8.0110180e-01,  6.5307493e+00, ...,\n",
       "         6.2219801e+00, -6.9380894e+00, -1.0412694e+00],\n",
       "       [ 5.4845099e+00,  1.9520372e-01,  4.7072082e+00, ...,\n",
       "         4.1869702e+00, -4.9142218e+00, -1.0798045e+00]], dtype=float32)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\n",
    "model = TranscriptomePredictor(CONFIG, GENE_FEAT_DIM, pathway_feats, chr_idx, locus_fourier).to(device)\n",
    "\n",
    "dataset = PerturbationDataset(adata, CONFIG) # Pass CONFIG to the dataset constructor\n",
    "train_loader = DataLoader(dataset, batch_size=CONFIG[\"batch_size\"], shuffle=True)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
